Fix frozen mutation consume ordering

bhalevy opened the pull request:
Currently, frozen_mutation is not consumed in position_in_partition
order as all range tombstones are consumed before all rows.
    
This violates the range_tombstone_generator invariants
as its lower_bound needs to be monotonically increasing.
    
Fix this by adding mutation_partition_view::accept_ordered
and rewriting do_accept_gently to do the same,
both making sure to consume the range tombstones
and clustering rows in position_in_partition order,
similar to the mutation consume_clustering_fragments function.
    
Add a unit test that verifies that.
    
Fixes #11198

avikivity commented:
Does the bug have any impact? Who are the consumers?

bhalevy commented:
> Does the bug have any impact? Who are the consumers?

The bug is manifested in the reconciliation path (`to_data_query_result`):
https://github.com/scylladb/scylladb/blob/8ee5b69f800481e31ecefbeae74d224d3d082152/mutation_partition.cc#L2213

I think that since each replica is compacting rows on its side we get only live rows and so the reordering doesn't create an issue for the query result builder.

tgrabiec commented:
to_data_query_result() is also called after reconciliation on coordinator side. The input frozen_mutation there may be the result of merging results from two replicas, and is not compacted. In this case, the compaction inside to_data_query_result() is needed to apply range tombstones across replica results.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1769/

denesb commented on frozen_mutation.hh:
```diff
@@ -51,20 +51,20 @@ private:
             auto row = std::move(_static_row.get_existing());
-            if (!_stop_consuming) {
-                _stop_consuming = _consumer.consume(static_row(std::move(row)));
+            if (_consumer.consume(static_row(std::move(row)))) {
```
I think producers shouldn't assume why a consumer wants to stop and shouldn't override this decision.

bhalevy replied:
Hmm, I think that can be done safely with the `frozen_mutation_consumer_adaptor` keeping the state, similar to the mutation::consume cookie.

bhalevy replied:
Also, the consumer must consume the range tombstone change after the last consumed row, even if it returned stop_iteration::yes for it.

denesb replied:
Yes, the producer has to enforce basic correctness on the consumer, but beyond that, if the consumer requests stop, the producer should stop as soon as "legally" allowed, with no assumptions about what consumers "really want".

bhalevy replied:
Fixed in v2

denesb commented on mutation_partition_view.cc:
```diff
@@ -230,11 +230,107 @@ void mutation_partition_view::do_accept(const column_mapping& cm, Visitor& visit
+                cr = *cr_it;
+                ++cr_it;
+                auto ck = cr->key();
+                if (ck == clustering_key(std::vector<bytes>{})) {
```
Use `is_empty()` instead, building an empty key is not free.

bhalevy replied:
Fixed in v2

denesb commented on mutation_partition_view.cc:
```diff
@@ -230,11 +230,107 @@ void mutation_partition_view::do_accept(const column_mapping& cm, Visitor& visit
+
+    std::optional<range_tombstone> rt;
+    auto next_rt = [&] {
+        if (!rt) {
```
```
if (rt) {
   return;
}
```
Would result in less indentation levels. Same below.

bhalevy replied:
Fixed in v2

denesb commented on mutation_partition_view.cc:
```diff
@@ -230,11 +230,107 @@ void mutation_partition_view::do_accept(const column_mapping& cm, Visitor& visit
 template<typename Visitor>
 requires MutationViewVisitor<Visitor>
-future<> mutation_partition_view::do_accept_gently(const column_mapping& cm, Visitor& visitor) const {
+future<> mutation_partition_view::do_accept_gently(const schema& s, Visitor& visitor) const {
```
If this method becomes ordered too, what is the point in `accept_ordered()`?

bhalevy replied:
accept_ordered is synchronous.
I'll see if they can be unified using a common synchronous do_accept_ordered function that get a `is_preemptible` parameter.

bhalevy replied:
Unified in v2

denesb commented on test/boost/frozen_mutation_test.cc:
```diff
@@ -169,3 +170,19 @@ SEASTAR_TEST_CASE(test_deserialization_using_wrong_schema_throws) {
+    random_mutation_generator gen{random_mutation_generator::generate_counters::no};
+    mutation m = gen();
+    auto& s = m.schema();
+    frozen_mutation fm{m};
```
Please also test with `mutation`s resulting from adding several different mutations together.

bhalevy replied:
What do you mean by "adding" mutation together?

denesb replied:
```
auto m1 = gen();
auto m2 = gen();
auto m = m1 + m2;
```
I don't think this will find anything but just to be sure...

bhalevy replied:
Done in v2

bhalevy commented:
In v2:
- frozen_mutation: consume and consume_gently in-order
  - templatized do_accept_ordered with <bool is_preemptible> parameter so it could be called from both `consume` and `consume_gently` paths
  - simplified `next_{rt,cr}` lambda functions as suggested
  - used cookie for preemptability, similar as mutation::consume
  - simplified consume loop + added optional preemptio
  - beefed-up test to include adding 2 random mutations
    - + compared consumed result to the original mutation using mutation_rebuilder_v2
- mutation_partition_view: make mutation_partition_view_virtual_visitor stoppable
  - added support for returning `stop_iteration` from consumer

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1826/

bhalevy commented:
In v3 (f2dfb267e0):
- frozen_mutation: consume and consume_gently in-order
  - Fixed bug causing skipping of rows with empty clustering key
    - copy-pasted form mutation consume skip of dummy rows that's not applicable to frozen mutations

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1836/

bhalevy commented:
@tgrabiec / @denesb can you please review?

bhalevy commented:
ping @tgrabiec / @denesb

denesb commented on frozen_mutation.hh:
```diff
@@ -290,7 +290,7 @@ auto frozen_mutation::consume(schema_ptr s, frozen_mutation_consumer_adaptor<Con
     try {
         adaptor.on_new_partition(_pk);
-        partition().accept(s->get_column_mapping(), adaptor);
+        partition().accept_ordered(*s, adaptor);
```
Why can't we leave the previous name? Do we want to keep an unordered consume variant too? For what? Won't that lead to mistakes?

bhalevy replied:
We have a few variants of `accept`, see https://github.com/scylladb/scylladb/blob/60e8f5743cc777882c6b53fa04a0e82c8ae862b2/mutation_partition_view.hh#L70-L75
The others may be used e.g. to unfreeze the mutation, and that doesn't need to be consumed in ordered since the mutation rebuilder reorders the range tombstones and rows when they are accepted into the mutation.
I think that doing that in order might hurt performance.

tgrabiec replied:
We could rename accept() to accept_unordered() to reduce probability of mistakes.

denesb commented on mutation_partition_view.cc:
```diff
@@ -276,6 +276,122 @@ future<> mutation_partition_view::do_accept_gently(const column_mapping& cm, Vis
+    const column_mapping& cm = s.get_column_mapping();
+
+    if (!cookie.accepted_partition_tombstone) {
+        mpvlog.trace("do_accept_ordered: accept partition tombstone: {}", tombstone(mpv.tomb()));
```
Do we really need logging here? I don't think we have it in any other producers.

bhalevy replied:
I needed it for development.  I presumed that the runtime penalty is negligible if the log level isn't high enough, so nothing gets computed.

denesb commented on mutation_partition_view.cc:
```diff
@@ -276,6 +276,122 @@ future<> mutation_partition_view::do_accept_gently(const column_mapping& cm, Vis
+            mutation_partition_view_virtual_visitor& _visitor;
+
+            void accept_atomic_cell(column_id id, atomic_cell ac) const {
+            _visitor.accept_static_cell(id, std::move(ac));
```
ident off

denesb commented on mutation_partition_view.cc:
```diff
@@ -276,6 +276,122 @@ future<> mutation_partition_view::do_accept_gently(const column_mapping& cm, Vis
+            _visitor.accept_static_cell(id, std::move(ac));
+            }
+            void accept_collection(column_id id, const collection_mutation& cm) const {
+            _visitor.accept_static_cell(id, cm);
```
same here

bhalevy commented:
In v4 (357e805e1f): removed logger, fixed indentation.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1894/

tgrabiec commented on frozen_mutation.hh:
```diff
@@ -51,9 +51,7 @@ private:
-            if (!_stop_consuming) {
-                _stop_consuming = _consumer.consume(static_row(std::move(row)));
-            }
+            _stop_consuming = _consumer.consume(static_row(std::move(row)));
```
It's strange that consumer can't stop before entering a partition.

tgrabiec commented on frozen_mutation.hh:
```diff
@@ -48,23 +48,26 @@ private:
-            if (!_stop_consuming) {
-                _stop_consuming = _consumer.consume(std::move(rt));
-            }
+            _stop_consuming = _consumer.consume(std::move(rt));
```
After this change, _consumer.consume() may be called after consume() returned stop_iteration::yes. Isn't it a regression?

tgrabiec commented on frozen_mutation.hh:
```diff
@@ -102,22 +102,22 @@ public:
         flush_rows_and_tombstones(rt.position());
         _rt_gen.consume(std::move(rt));
-        // FIXME: propagate stop_consuming to caller
+        return _stop_consuming;
```
In one place you check the result of flush_rows_and_tombstones(), here you check _stop_consuming. Perhaps we should stick to one method.
